src/merkle/merkle_mountain_range/full.rs

```diff
@@ -0,0 +1,356 @@
+/// the `count` is not a power-of-two or zero.
+fn num_of_elements_to_tree_size(count: usize) -> usize {
+    assert!(
+        count.is_power_of_two(),
+        "count must be a power-of-two {count}"
+    );
+
+    (count << 1) - 1
+}
+
+/// Iterate over the bits of a `usize` and yields the bit positions for the true bits.
+struct BitPositionIterator {
+    value: usize,
+}
+
+impl BitPositionIterator {
+    fn new(value: usize) -> BitPositionIterator {
+        BitPositionIterator { value }
+    }
+}
+
+impl Iterator for BitPositionIterator {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<<Self as Iterator>::Item> {
+        // trailing_zeros is compute with the intrinsic cttz. [Rust 1.67.0] x86 uses the `bsf`
+        // instruction. AArch64 uses the `rbit clz` instructions.
+        let zeros = self.value.trailing_zeros();
+
+        if zeros == usize::BITS {
+            None
+        } else {
+            let bit_position = zeros;
+            let mask = 1 << bit_position;
+            self.value ^= mask;
+            Some(bit_position)
+        }
+    }
+}
+
+impl DoubleEndedIterator for BitPositionIterator {
+    fn next_back(&mut self) -> Option<<Self as Iterator>::Item> {
+        // trailing_zeros is compute with the intrinsic ctlz. [Rust 1.67.0] x86 uses the `bsr`
+        // instruction. AArch64 uses the `clz` instruction.
+        let zeros = self.value.leading_zeros();
+
+        if zeros == usize::BITS {
+            None
+        } else {
+            let bit_position = usize::BITS - zeros - 1;
+            let mask = 1 << bit_position;
+            self.value ^= mask;
+            Some(bit_position)
+        }
+    }
+}
+
+pub struct Accumulator {
+    /// The number of leaves is used to differentiate accumulators that have the same number of
+    /// peaks. This happens because the number of peaks goes up-and-down as the structure is used
+    /// causing existing trees to be merged and new ones to be created. As an example, every time
+    /// the MMR has a power-of-two number of leaves there is a single peak.
+    ///
+    /// Every tree in the MMR forest has a distinct power-of-two size, this means only the right
+    /// most tree can have an odd number of elements (1). Additionally this means that the bits in
+    /// `leaves` conveniently encode the size of each individual tree.
+    ///
+    /// Examples:
+    ///
+    ///    Example 1: With 5 leaves, the binary 0b101. The number of set bits is equal the number
+    ///    of peaks, in this case there are 2 peaks. The 0-indexed least-significant position of
+    ///    the bit determines the number of elements of a tree, so the rightmost tree has 2**0
+    ///    elements and the left most has 2**2.
+    ///
+    ///    Example 2: With 12 leaves, the binary is 0b1100, this case also has 2 peaks, the
+    ///    leftmost tree has 2**3=8 elements, and the right most has 2**2=4 elements.
+    pub leaves: usize,
```

Comment:
I only read the the paper [1] and grin's documentation [2]. I think using the bits of `leaves` is a much cleaner way of figuring which tree peaks we can merge, since that doesn't need any search, just iterate over the bits of `leaves` while they are set, starting from the least-significant until the first `0` bit. This also neatly maps to the postorder sequence representation I used to avoid moving elements around.


1- Efficient Asynchronous Accumulators for Distributed PKI [ref](https://eprint.iacr.org/2015/718.pdf)
2- https://docs.grin.mw/wiki/chain-state/merkle-mountain-range/